Source/Events.Handling/Builder/ConventionEventHandlerBuilder.cs

```diff
@@ -0,0 +1,319 @@
+            {
+                buildResult = new BuildEventHandlerResult(
+                    eventHandlerId,
+                    $"There are no event handler methods to register in event handler {_eventHandlerType}. An event handler method either needs to be decorated with [{typeof(HandlesAttribute).Name}] or have the name {MethodName}");
+                return false;
+            }
+
+            buildResult = new BuildEventHandlerResult();
+            return true;
+        }
+
+        bool TryAddDecoratedHandlerMethods(
+            IEnumerable<MethodInfo> methods,
+            EventHandlerId eventHandlerId,
+            EventHandlerMethodsBuilder eventHandlerMethodsBuilder,
+            out EventHandlerBuildWarnings warnings)
+        {
+            warnings = default;
+            var warningMessages = new List<string>();
+            foreach (var method in methods.Where(IsDecoratedHandlerMethod).ToArray())
+            {
+                var shouldAddHandler = true;
+                var eventType = (method.GetCustomAttributes(typeof(HandlesAttribute), true)[0] as HandlesAttribute)?.EventType;
+                if (!TryGetFirstMethodParameterType(method, out var eventParameterType))
+                {
+                    warningMessages.Add($"Event handler method {method} on event handler {_eventHandlerType} has no parameters, but is decorated with [{typeof(HandlesAttribute).Name}]. An event handler method should take in as paramters an event and an {typeof(EventContext).Name}");
+                    shouldAddHandler = false;
+                }
+
+                if (!ParametersAreOkay(method, eventHandlerId, out var parameterWarnings))
+                {
+                    warningMessages.AddRange(parameterWarnings.Warnings);
+                    shouldAddHandler = false;
+                }
+
+                if (eventParameterType != typeof(object))
+                {
+                    warningMessages.Add($"Event handler method {method} on event handler {_eventHandlerType} should only handle an event of type object");
+                    shouldAddHandler = false;
+                }
+
+                if (shouldAddHandler) AddUntypedHandleSignatureToBuilder(method, eventType, eventHandlerMethodsBuilder);
+            }
+
+            if (warningMessages.Count == 0) return true;
+
+            warnings = new EventHandlerBuildWarnings(eventHandlerId, warningMessages);
+            return false;
+        }
+
+        bool TryAddConventionHandlerMethods(
+            IEnumerable<MethodInfo> methods,
+            EventHandlerId eventHandlerId,
+            IEventTypes eventTypes,
+            EventHandlerMethodsBuilder eventHandlerMethodsBuilder,
+            out EventHandlerBuildWarnings warnings)
+        {
+            warnings = default;
+            var warningMessages = new List<string>();
+            foreach (var method in methods.Where(_ => !IsDecoratedHandlerMethod(_) && _.Name == MethodName).ToArray())
```

Comment:
`.ToArray()`?

Reply:
Lazy evaluation. I have burned myself on that before